serve/mlc_serve/engine/sync_engine.py

```diff
@@ -273,15 +279,21 @@ def _adjust_batch(self):
                 state = self.queue[0]
                 num_tokens = len(state.token_ids)
                 num_new_batched_tokens += num_tokens
+                # this can happen if we processed some request and then had to remove it from inference
+                # due to limit of the kv cache. It can appear that new size to process is bigger
+                # than max_num_batched_tokens. Need to roll back to acceptable size
+                if not len(self.current_batch) and num_new_batched_tokens > self.max_num_batched_tokens:
+                    state.token_ids = state.token_ids[:self.max_num_batched_tokens]
+                    state.next_start_position = num_new_batched_tokens = num_tokens = self.max_num_batched_tokens
                 if num_new_batched_tokens > self.max_num_batched_tokens > 0:
                     logger.debug(
                         "Stop growing the batch due to max_num_batched_tokens. Batched tokens: %s",
                         num_new_batched_tokens,
                     )
                     break
                 if (
-                    self.cache_manager.get_free_space()
-                    <= self.prompt_allocate_ratio * num_tokens
+                    (self.cache_manager.get_free_space() - num_tokens) / (len(self.current_batch) + 1)
+                        < self.max_decode_steps
```

Comment:
For the change in sync engine, please apply my comments on the staging engine to here as well. Or extract common conditions into a utility that can be used by both engines.

Reply:
> Or extract common conditions into a utility that can be used by both engines.

It's too big refactoring, let's plan it as a separate PR. In other case this PR will never landed )